gh-125997: ensure that `time.sleep(0)` is not delayed on non-Windows platforms

picnixz opened the pull request:
This is a suggestion for fixing the issue when the timeout is 0. For other timeouts, this does not change the current behaviour (I'm tempted removing the `do-while` loop as we should never have an issue, but I'm not familar with the monotonic C implementation so I left it as is). 

<!-- gh-issue-number: gh-125997 -->
* Issue: gh-125997
<!-- /gh-issue-number -->

picnixz committed:
blurb

picnixz committed:
Ensure that `time.sleep(0)` does not accumulate delays.

On non-Windows platforms, this reverts the usage of `clock_nanosleep`
and `nanosleep` introduced by 85a4748 and 7834ff2 respectively,
falling back to a `select(2)` alternative instead.

picnixz committed:
update globals

picnixz committed:
fix Windows compilation

rruuaanng commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2398,44 @@ pysleep(PyTime_t timeout)
+
+
+#ifndef MS_WINDOWS
+// time.sleep(0) optimized implementation.
+// On error, raise an exception and return -1.
+// On success, return 0.
```
```suggestion
/** 
  * time.sleep(0) optimized implementation.
  * On error, raise an exception and return -1.
  * On success, return 0.
  */
```
It is recommended to use C89 comments as function descriptions.
(Maybe this is a personal preference, but I still recommend it.)

picnixz replied:
I'm aligning it with the comments of `time.sleep()`.

picnixz commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2398,44 @@ pysleep(PyTime_t timeout)
+    do {
+        int ret, err;
+        Py_BEGIN_ALLOW_THREADS
+        ret = select(0, (fd_set *)0, (fd_set *)0, (fd_set *)0, &zero);
```
FTR: [`select(2)`](https://man7.org/linux/man-pages/man2/select.2.html) can mutate the timeout argument but if both fields are zero it immediately returns.

ZeroIntensity commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2398,44 @@ pysleep(PyTime_t timeout)
+            break;
+        }
+        if (err != EINTR) {
+            errno = err;
```
It might be worth asserting that the function is called without an errno set.

picnixz replied:
~~Considering it's at the very bottom of the file and that it's really an implementation detail, I wouldn't bother about it. I'm not sure the existing code checked it. But I'll add an assert at the beginning.~~

See https://github.com/python/cpython/pull/128274#issuecomment-2563593057.

ZeroIntensity commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2398,44 @@ pysleep(PyTime_t timeout)
+static int
+pysleep_zero_posix(void)
+{
+    static struct timeval zero = {0, 0};
```
Hmm, why make this a static local instead of a const global?

picnixz replied:
`select(2)` does not allow for a constant since it can modify the pointer (but it wouldn't in practice since both fields are 0).

ZeroIntensity replied:
I'm not sure I trust that for all systems, we might end up with some interesting edge cases. It's probably safer to just make it a thread local global.

picnixz replied:
I'll just make a local variable simply. That will reduce the diff and avoid bothering about portability (it eventually shouldn't really matter performance-wise probably)

picnixz committed:
use a fresh `timeval` variable for select(2) portability

picnixz committed:
ensure that `errno=0` when calling `pysleep[zero_posix]`

picnixz committed:
removing `assert(errno == 0)`

Due to how `OSError` are raised from `errno`, we do not clear `errno`
afterwards. If we catch `OSError`, then we still have an errno set,
and if we call `time.sleep()` just after, we may have `errno != 0`
(but we know we handled it so it's fine).

picnixz committed:
add errors assertions

hauntsaninja commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2399,53 @@ pysleep(PyTime_t timeout)
+        struct timeval zero = {0, 0};
+        Py_BEGIN_ALLOW_THREADS
+        ret = select(0, (fd_set *)0, (fd_set *)0, (fd_set *)0, &zero);
+        err = errno;
```
nit: I think we can get rid of `err`, seems like an artifact of the version of this code that has a bunch of #ifdef

hauntsaninja reviewed (commented):
Linking https://github.com/python/cpython/pull/28111 for other reviewers

I'd vote we can get rid of the do-while, I think it would only matter if monotonic is not monotonic

I'd also vote we don't backport. This is basically a performance optimisation (which we don't typically backport) for a regression that has only been reported once in two years and that has an easy workaround.

picnixz committed:
simplify code

hauntsaninja reviewed (commented):
I think removing them is fine. This PR lgtm! I'm not sure I understand what you mean by advance, what behaviour are you worried about?

It also looks like `test_time.py` doesn't have a `time.sleep(0)` (there aren't many `time.sleep(0)` in CPython) — would you mind adding a very basic test?

hauntsaninja reviewed (commented):
Some more thoughts:
- Since this behaviour is specifically circumventing the default Linux scheduler's behaviour, should we only do this on Linux?
    - Note empirically it looks like the default Linux scheduler affects all short sleeps (under 50us), not just zero sleep. I don't want CPython to get into the business of trying to detect that. Maybe users who care about this should be doing `os.sched_setscheduler`? (edit: looks like you could use `PR_SET_TIMERSLACK` too)
- Maybe `sched_yield` is a more natural thing than `select`? What else do users want `time.sleep(0)` for anyway?
    - I can't find anything about `select` that doesn't look like an implementation detail, which I guess makes me more inclined to `sched_yield`
    - Using `sched_yield` might make doing this on all POSIX not just Linux more principled
- If we merge a special case for zero sleep, we should document in code why we do it

cc @charles-cooper in case you have opinions, would love to know more about your use case.

picnixz committed:
add comments for `time.sleep(0)`

picnixz committed:
improve time.sleep() tests

picnixz committed:
update docs

picnixz committed:
add What's New

picnixz committed:
update `os` scheduling policy docs

picnixz committed:
fix typo

picnixz committed:
optimize `time.sleep(0)` path

picnixz committed:
update docs

picnixz committed:
update tests

picnixz committed:
revert un-necessary docs updates

vstinner commented on Lib/test/test_time.py:
```diff
@@ -158,10 +158,45 @@ def test_conversions(self):
+                time.sleep(value)
+
+    @unittest.skipIf(support.MS_WINDOWS, 'test only for non-Windows platforms')
+    def test_sleep_zero_posix(self):
```
I dislike "performance" test. They tend to always fail on slowest buildbots.

picnixz replied:
I can totally remove it. I think it was more important to have the type and value tests.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2399,60 @@ pysleep(PyTime_t timeout)
+// skip some calls to `PyTime_Monotonic()` and other checks when the timeout
+// is zero. For details, see https://github.com/python/cpython/pull/128274.
+static int
+pysleep_zero_posix(void)
```
Can you write a single pysleep_zero() function for all platforms, and have a code path for Windows? Move this code:

```
        Py_BEGIN_ALLOW_THREADS
        // A value of zero causes the thread to relinquish the remainder of its
        // time slice to any other thread that is ready to run. If there are no
        // other threads ready to run, the function returns immediately, and
        // the thread continues execution.
        Sleep(0);
        Py_END_ALLOW_THREADS
```

vstinner commented on Doc/whatsnew/3.14.rst:
```diff
@@ -617,6 +617,15 @@ sys.monitoring
+time
+----
+
+* Ensure that :func:`time.sleep(0) <time.sleep>` does not degrade over time
```
I suppose that you're referring to performance:

```suggestion
* Ensure that :func:`time.sleep(0) <time.sleep>` performance does not degrade over time
```

picnixz replied:
Depending on the resolution of this PR (namely, whether we eventually use `select()`), I'll amend the NEWS so I won't commit that suggestion for now.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2399,60 @@ pysleep(PyTime_t timeout)
+    err = ret;
+#elif defined(HAVE_NANOSLEEP)
+    struct timespec zero = {0, 0};
+    ret = nanosleep(&zero, NULL);
```
Is it worth it to have 3 code paths for sleep(0)? Can't we always use select()?

picnixz replied:
We had a long discussion on this matter but long story short:

- `select()` is implementation sensitive but there were some semantical and implementation concerns (see https://github.com/python/cpython/pull/128274#issuecomment-2565291149).
- using `select()` would be much faster and if you still want to go that road (which I did originally), then I personally don't mind.

picnixz committed:
remove performance tests

picnixz committed:
implement platform compile-time dependent `time.sleep(0)`

The implementations of `time.sleep(0)` on Windows and POSIX are
now handled by the same function. Previously, `time.sleep(0)` on
Windows was handled by `pysleep()` while on POSIX platforms, it
was handled by `pysleep_zero_posix()`.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2213,6 +2214,19 @@ static int
     assert(timeout >= 0);
+    assert(!PyErr_Occurred());
+#ifndef MS_WINDOWS
+    if (timeout == 0) { // gh-125997
```
This test also works on Windows. `_PyTime_As100Nanoseconds(timeout, _PyTime_ROUND_CEILING)` can only return zero is timeout is zero.

picnixz replied:
Actually, it would but `_PyTime_As100Nanoseconds` is only defined on Windows:

```c
#ifdef MS_WINDOWS
PyTime_t
_PyTime_As100Nanoseconds(PyTime_t ns, _PyTime_round_t round)
{
    return pytime_divide(ns, NS_TO_100NS, round);
}
#endif
```

Should I undef it?

vstinner replied:
I'm suggesting to use `if (timeout == 0)` test on Windows, and keep the _PyTime_As100Nanoseconds() code for code above.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2389,67 @@ pysleep(PyTime_t timeout)
+    if (PyErr_CheckSignals()) {
+        return -1;
+    }
+#else
```
You may add:

```
// Windows implementation
```

vstinner commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2389,67 @@ pysleep(PyTime_t timeout)
+//
+// Rationale
+// ---------
+// time.sleep(0) accumulates delays in the generic implementation, but we can
```
"accumulates delays" I don't understand this part.

picnixz replied:
Ah right, `time.sleep(0)` itself is not accumulating delays. I'll rewrite that part tomorrow or during the week.

picnixz committed:
address Victor's review

picnixz committed:
update blurb

picnixz committed:
simplify `pysleep`

vstinner commented on Modules/timemodule.c:
```diff
@@ -2390,3 +2382,67 @@ pysleep(PyTime_t timeout)
+    Py_BEGIN_ALLOW_THREADS
+#ifdef HAVE_CLOCK_NANOSLEEP
+    struct timespec zero = {0, 0};
+    ret = clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &zero, NULL);
```
I don't think that TIMER_ABSTIME is appropriate here:

```suggestion
    ret = clock_nanosleep(CLOCK_MONOTONIC, 0, &zero, NULL);
```

picnixz replied:
I first thought about:

> If flags is TIMER_ABSTIME, then t is interpreted as an absolute
  time as measured by the clock, clockid. If t is less than or
  equal to the current value of the clock, then clock_nanosleep()
  returns immediately without suspending the calling thread.

Without this flag, `time.sleep(0)` takes 50us. Otherwise it takes 2us. So, now I'm more and more inclined to actually revert it back to a `select`. Because otherwise, it's as if I'm not doing anything (and just skip the call to `clock_nanosleep`; I mean I already know that the condition on `t` is verified so it's as if I have no call)

picnixz replied:
I'm leaving for a few days but I'm still struggling to convince myself between the use of `select` or not. The reason why `clock_nanosleep()` is slowed down is because we don't pass a zero time struct.

picnixz replied:
Ok, I was wrong. See https://github.com/python/cpython/pull/128274#issuecomment-2575513339.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2213,7 +2214,10 @@ static int
+    assert(!PyErr_Occurred());
+    if (timeout == 0) { // gh-125997
+        return pysleep_zero();
+    }
```
Empty line for readability:

```suggestion
    }

```

picnixz committed:
add newline for readability

picnixz committed:
revert to a `select()` alternative

Using `clock_nanosleep()` would always take more than 50 us.
